Full review (DT-17)

DaveDangereux opened the pull request:
<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

- **New Features**
	- Introduced a new puzzle game named `ball_sort`, involving sorting colored balls into tubes.
	- Added functionality to create puzzles, solve them using a solver, and interactively play the solution.
	- Implemented a `Move` structure to represent moves within the game, including origin, destination, and serialized state.
	- Developed classes for managing game elements: `Puzzle` for game logic, `Tube` for tube and ball interactions, and `Solver` for solving puzzles.
- **Documentation**
	- Updated README with non-specific changes.
- **Chores**
	- Set up project structure and compilation options using CMake.
	- Provided a script for building, cleaning, and managing the project.
- **Tests**
	- Added infrastructure to set up and run tests using the Google Test framework.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

DaveDangereux committed:
Add initial project files

DaveDangereux committed:
Implement get_legal_moves in Puzzle class

DaveDangereux committed:
Implement basic brute-force solver

DaveDangereux committed:
Add gtest

DaveDangereux committed:
Add solution playback, useless move check and strict diagnostics

DaveDangereux committed:
Move compile options to executable target

DaveDangereux committed:
Extract methods and reorganise for clarity

DaveDangereux committed:
Move source files to subdirectory

DaveDangereux committed:
Rename main class

DaveDangereux committed:
Extract puzzle data and solver into separate classes

DaveDangereux committed:
Make terminal clear command Windows compatible

DaveDangereux committed:
Move all state from solver to puzzle

DaveDangereux committed:
Divide move validation and move selection between puzzle and solver

DaveDangereux committed:
Tidying and renaming for maximum comfort

DaveDangereux committed:
Apply more thorough const referencing and erase redundant operator overloads

DaveDangereux committed:
Replace Makefile of personal aliases with bash scripts

DaveDangereux committed:
Fix bug with multiple nodemon instances restarting each other

DaveDangereux committed:
Remove artefacts from personal dev environment

DaveDangereux committed:
Remove all references to nodemon and tmux from scripts

DaveDangereux committed:
Add default case to case statement in scripts.sh

D4n93r3ux commented:
@coderabbitai review

JustCallMeRay commented on CMakeLists.txt:
```diff
@@ -0,0 +1,17 @@
+
+include_directories(${CMAKE_SOURCE_DIR}/include)
+
+add_compile_options(-Wall -Wextra -Wshadow -Wconversion -Wpedantic -Werror)
```
The compiler flags are different for most compilers, wrap this in an if statement so it works for MSVC also best to default to off if not gcc/clang so if people are using their weird things it still works, but not a big deal.

JustCallMeRay commented on CMakeLists.txt:
```diff
@@ -0,0 +1,17 @@
+add_compile_options(-Wall -Wextra -Wshadow -Wconversion -Wpedantic -Werror)
+
+add_subdirectory(app)
+add_subdirectory(tests)
```
Normally you want to have an option to not compile the tests so users can avoid excess processing/compiling 

At work compiling tests for one specific thing takes about 3x as compiling the thing it's self

JustCallMeRay commented on app/CMakeLists.txt:
```diff
@@ -0,0 +1,8 @@
+target_include_directories(${PROJECT_NAME} PRIVATE ${CMAKE_SOURCE_DIR}/include)
+
+target_compile_options(${PROJECT_NAME} PRIVATE -Wall -Wextra -Wshadow
+                                               -Wconversion -Wpedantic -Werror)
```
Again the if statement

JustCallMeRay commented on app/app.cpp:
```diff
@@ -0,0 +1,17 @@
+int main()
+{
+    Puzzle puzzle{"9 10 1 5 2 12 8 12 10 4 6 10 8 5 7 1 10 11 3 7 5 4 1 2 3 8 "
+                  "1 9 2 7 12 8 4 9 11 6 6 12 2 3 6 3 11 9 5 4 7 11"};
```
If this can be randomised and the algorithm still works I would suggest that.

Also this is PERFECT for fuzz testing, do everything else first though!

DaveDangereux committed:
Remove redundant forward declaration

JustCallMeRay commented on app/app.cpp:
```diff
@@ -0,0 +1,17 @@
+
+    Solver::solve(puzzle);
+
+    std::cin.get();
```
If this is just used as a pause print some text out to the user to make it more obvious.

JustCallMeRay commented on app/app.cpp:
```diff
@@ -0,0 +1,17 @@
+
+    Solver::play_solution(puzzle);
+
+    return 0;
```
Return value is unneeded in main but feel free to keep it in (it's a stylistic choice, I have no string opinions about it)

(You may have seen it in the code I sent you and that was more because my IDE does not have a good way to compile one file into one executable, similar to cmake)

JustCallMeRay commented on bin:
```diff
@@ -0,0 +1 @@
+./build/bin
```
I think this whole bin symlink thing is pointless and odd. You can (probably) redirect where the output of the compiled code goes in the cmake if you want a top level bin folder

JustCallMeRay commented on include/ball_sort/move.hpp:
```diff
@@ -0,0 +1,29 @@
+struct std::hash<Move> {
+    auto operator()(const Move &move) const -> size_t
+    {
+        return (std::hash<size_t>()(move.m_origin) << 2) ^
+               std::hash<size_t>()(move.m_destination) ^
+               (std::hash<std::string>()(move.m_serialised_state) << 4);
```
I'd prefer a `std::tie` or a `std::make_tuple` then hash that (assuming it works in GCC, I've only tried in msvc)

DaveDangereux replied:
Wouldn't I then have to specialise std::hash for the tuple? Which leaves me in an equivalent position?

Demo code would be lovely.

JustCallMeRay commented on include/ball_sort/move.hpp:
```diff
@@ -0,0 +1,29 @@
+    const size_t m_destination;
+
+ private:
+    const std::string m_serialised_state;
```
default this here instead of in the constructor.

Here specifically it doesn't matter but generally you can often default the `Type()` constructor if all the variables are defaulted where declared (so if you add one in the future it's also a smaller change though it doesn't make sense to have one here)

JustCallMeRay commented on include/ball_sort/puzzle.hpp:
```diff
@@ -0,0 +1,39 @@
+
+class Puzzle {
+ public:
+    Puzzle(const std::vector<std::string> &letter_strings);
```
I'd take an auto here so I can pass in an array, (allowing constexpr in C++17), list, set etc. 

Source code doesn't have to change (well types obvs but that's it). Consider a concept too!

JustCallMeRay commented on include/ball_sort/puzzle.hpp:
```diff
@@ -0,0 +1,39 @@
+class Puzzle {
+ public:
+    Puzzle(const std::vector<std::string> &letter_strings);
+    Puzzle(const std::string &number_sequence);
```
Think about taking a stringview here

JustCallMeRay commented on include/ball_sort/puzzle.hpp:
```diff
@@ -0,0 +1,39 @@
+    auto is_solved() const -> bool;
+    auto reset() -> void;
+    auto get_tubes() const -> const std::vector<Tube> &;
+    auto get_history() const -> const std::vector<Move> &;
```
Everything non-void here should be `[[nodiscard]]`

JustCallMeRay commented on include/ball_sort/puzzle.hpp:
```diff
@@ -0,0 +1,39 @@
+    auto is_legal_move(const size_t origin_index,
+                       const size_t destination_index) const -> bool;
+    auto get_ball_tally() const -> std::unordered_map<char, size_t>;
+    constexpr auto get_serialised_state() const -> std::string;
```
Again [[nodiscard]]

JustCallMeRay commented on CMakeLists.txt:
```diff
@@ -0,0 +1,17 @@
+
+project(ball_sort LANGUAGES CXX)
+
+set(CMAKE_CXX_STANDARD 20)
```
If you were making a lib this would need to be in an if block so the user could specify a higher version.

Don't do it here though

JustCallMeRay commented on include/ball_sort/solver.hpp:
```diff
@@ -0,0 +1,16 @@
+    static auto pick_move(const std::vector<Move> &filtered_moves) -> Move;
+    static auto print_puzzle(const Puzzle &puzzle) -> void;
+    static auto play_solution(Puzzle &puzzle) -> void;
+};
```
I generally only have static classes for compile time dependency injection (I'll write you a file, might have one) and generally prefer functions and a namespace for this.

It's stylistic so up to you.

Also think I'd have a normal class that owns a puzzle? I'm not sure

JustCallMeRay commented on include/ball_sort/tube.hpp:
```diff
@@ -0,0 +1,27 @@
+
+class Tube {
+ public:
+    Tube(const std::string &balls);
```
Consider taking a stringview

JustCallMeRay commented on include/ball_sort/tube.hpp:
```diff
@@ -0,0 +1,27 @@
+    auto get_top_ball() const -> char;
+    auto get_serialised_balls() const -> std::string;
+    auto take_top_ball() -> char;
+    auto place_ball(const char ball) -> void;
```
[[nodiscard]]

JustCallMeRay commented on include/ball_sort/tube.hpp:
```diff
@@ -0,0 +1,27 @@
+    auto is_full() const -> bool;
+    auto is_one_colour() const -> bool;
+    auto is_solved() const -> bool;
+    auto get_balls() const -> const std::string &;
```
Who owns the string you are returning?

If this is the error I think this is address sanitizers will catch it

JustCallMeRay commented on include/ball_sort/tube.hpp:
```diff
@@ -0,0 +1,27 @@
+    auto take_top_ball() -> char;
+    auto place_ball(const char ball) -> void;
+
+    static const size_t MAX_CAPACITY{4};
```
Think this should be private

JustCallMeRay commented on scripts.sh:
Worst named file ever

I would separate it into multiple files over a single one and maybe put them in a folder

JustCallMeRay commented on scripts.sh:
```diff
@@ -0,0 +1,31 @@
+test_executable=ball_sort_tests
+
+if [ $# -eq 0 ]; then
+	echo "Usage: $0 <command>"
```
Say what the possible commands are, the user would rather not read shell script

JustCallMeRay commented on src/ball_sort/move.cpp:
```diff
@@ -0,0 +1,8 @@
+           const std::string &serialised_state)
+    : m_origin{origin}, m_destination{destination},
+      m_serialised_state{serialised_state}
+{}
```
Files like this I think I'd keep in the header unless I thought the file is going to get bigger later on.

Think this can also be contexpr which might make it more appealing

JustCallMeRay commented on src/ball_sort/puzzle.cpp:
```diff
@@ -0,0 +1,186 @@
+
+    for (const auto &ball_string : letter_strings) {
+        tubes.emplace_back(ball_string);
+    }
```
You are reinventing the ~~copy, just do `a = b`~~ transform.

`std::transform` with `std::execution::unseq` (or similar) will be faster

JustCallMeRay commented on src/ball_sort/puzzle.cpp:
```diff
@@ -0,0 +1,186 @@
+constexpr auto Puzzle::make_tubes(
+    const std::vector<std::string> &letter_strings) -> std::vector<Tube>
+{
+    std::vector<Tube> tubes{};
```
reserve the size of this vector

JustCallMeRay commented on src/ball_sort/puzzle.cpp:
```diff
@@ -0,0 +1,186 @@
+            tubes.emplace_back(ball_string);
+            ball_string.clear();
+        }
+    }
```
this feels odd, consider a refactor (I don't have any ideas though)

JustCallMeRay commented on src/ball_sort/puzzle.cpp:
```diff
@@ -0,0 +1,186 @@
+
+    while (number_stream >> number) {
+        ball_string.push_back(static_cast<char>('A' - 1 + number));
+        if (ball_string.size() == 4) {
```
4 is hard coded here and I think it's variable somewhere else (but not in the puzzle header?) have a look and check things work as expected. 
(assuming it's size of tube)

JustCallMeRay commented on src/ball_sort/puzzle.cpp:
```diff
@@ -0,0 +1,186 @@
+
+    if (!is_valid_puzzle) {
+        std::cout << "Invalid puzzle" << '\n';
+        exit(1);
```
NEVER call exit, throw an exception or take a callback to call instead.

I once worked with a dll that called exit and it was a nightmare, it is not a good habit to get into, it is rarely required.

Also I believe the new hotness is terminate (but no for the same reasons)

JustCallMeRay commented on src/ball_sort/puzzle.cpp:
```diff
@@ -0,0 +1,186 @@
+            if (is_legal_move(o, d))
+                legal_moves.emplace_back(o, d, current_state);
+        }
+    }
```
shortenings

JustCallMeRay commented on src/ball_sort/puzzle.cpp:
```diff
@@ -0,0 +1,186 @@
+    m_move_history.emplace_back(origin, destination, state_prior_to_move);
+
+    char ball{m_tubes.at(origin).take_top_ball()};
+    m_tubes.at(destination).place_ball(ball);
```
`at` is a little slower than `[ ]` and if you are using address sanitizers in testing they will crash the program instead of throwing. `[ ]` is generally preferred at work but I like `at` if I am doing constexpr stuff (I don't know why I think it just feels right, maybe no reason)

JustCallMeRay commented on src/ball_sort/puzzle.cpp:
```diff
@@ -0,0 +1,186 @@
+
+auto Puzzle::do_move(const size_t origin, const size_t destination) -> void
+{
+    const std::string &state_prior_to_move{get_serialised_state()};
```
Think this should be an owning string

JustCallMeRay commented on src/ball_sort/puzzle.cpp:
```diff
@@ -0,0 +1,186 @@
+    char ball{m_tubes.at(origin).take_top_ball()};
+    m_tubes.at(destination).place_ball(ball);
+
+    const std::string &current_state{get_serialised_state()};
```
Again owning?

JustCallMeRay commented on src/ball_sort/puzzle.cpp:
```diff
@@ -0,0 +1,186 @@
+    m_tubes = m_initial_state;
+    m_move_history.clear();
+    m_visited_puzzle_states.clear();
+}
```
When would you reset over making a new instance of the class?

DaveDangereux replied:
When calling solver::play_solution, the history (representing the solution - perhaps I should have called it that) is copied out and the puzzle is reset prior to playback, so as to not create a duplicate history (which would affect sequential playbacks). So the instance is reused.

It could probably do with a rethink, but presently I'm just trying to stay afloat in a sea of new things to learn whilst implementing as many of your suggestions as I can manage :]

I might rename that variable, though.

JustCallMeRay commented on src/ball_sort/puzzle.cpp:
```diff
@@ -0,0 +1,186 @@
+    const Tube &origin{m_tubes.at(origin_index)};
+    const Tube &destination{m_tubes.at(destination_index)};
+
+    if (origin_index == destination_index) return false;
```
Check before creating tube references (avoids possible exception and makes code quicker)

JustCallMeRay commented on src/ball_sort/puzzle.cpp:
```diff
@@ -0,0 +1,186 @@
+
+auto Puzzle::get_ball_tally() const -> std::unordered_map<char, size_t>
+{
+    std::unordered_map<char, size_t> ball_tally{};
```
Again I'd use a flat map like thing for speed

JustCallMeRay commented on src/ball_sort/puzzle.cpp:
```diff
@@ -0,0 +1,186 @@
+
+    for (const Tube &tube : m_tubes) {
+        for (const char ball : tube.get_balls()) {
+            ball_tally[ball]++;
```
Think this relies on the size_t being set to 0 when a new instance is created which might be undefined behaviour. 

If using clang a sanitizer might catch this but its worth checking.

DaveDangereux replied:
As far as I can tell, size_t will always initialise to 0 in this context. Sanitisers have not caught anything yet, at least.

JustCallMeRay commented on src/ball_sort/puzzle.cpp:
```diff
@@ -0,0 +1,186 @@
+
+constexpr auto Puzzle::get_serialised_state() const -> std::string
+{
+    std::string serialised_state{};
```
Think you can reserve the size of this string, at least to `m_tubes.size()` (though I forget how beneficial it is for strings)

D4n93r3ux replied:
std::string appears to default to 22 chars for me, so it seems that reserving would be unnecessary. Apparently this is called Short String Optimisation.

JustCallMeRay commented on src/ball_sort/solver.cpp:
```diff
@@ -0,0 +1,118 @@
+
+    print_puzzle(puzzle);
+
+    std::unordered_set<Move> excluded_moves{};
```
Feel like this would be quicker to have a vector. 

Consider C++23's flat set or write your own

DaveDangereux replied:
Seems like my version of Clang doesn't support flat_set.

I ran some tests in a separate sandbox with a vector and an unordered_set of 10000 items (reserved).

Things that I learnt:
- reserving is good
- vectors smash sets for emplacing
- std::find is better than std::unordered_set::contains, although I don't know exactly why yet
- sets smash vectors for lookup

Nothing particularly surprising there.

However, for some reason when I actually run the program with excluded_moves as a vector, it takes almost twice as long (120ms vs 70ms).

I assume this is because I'm doing more lookups than emplaces.

Reserving the size of excluded moves makes a fractional improvement to the vector performance.

Applying compiler optimisations makes them equivalent at around 30ms, though.

I'm inclined to go with the set for now, since it's faster without optimisation and equivalent with. But this has been a fun learning journey.

DaveDangereux replied:
Actually, std::unordered set is faster with optimisation too. 20ms vs 30ms for the vector.

That seals it.

DaveDangereux replied:
18ms with reserve. Woop.

JustCallMeRay replied:
I assume you are using the emplace back with a move rather than leaving it as an insert? 

Seems odd but I'll go with it :)

JustCallMeRay commented on include/ball_sort/solver.hpp:
```diff
@@ -0,0 +1,16 @@
+ public:
+    static auto solve(Puzzle &puzzle) -> void;
+    static auto
+    purge_redundant_moves(const std::vector<Move> &legal_moves,
```
Should this be private?

JustCallMeRay commented on src/ball_sort/solver.cpp:
```diff
@@ -0,0 +1,118 @@
+    return filtered_moves.front();
+}
+
+auto Solver::print_puzzle(const Puzzle &puzzle) -> void
```
I thought a little about this and how to make it better.

Refactor this to take a stream as an input and a callback to clear the stream, defaulting it to (do) nothing.
Create an overload which calls this with `std::cout` and `const auto func = []() { std::system("cls); };`

JustCallMeRay commented on src/ball_sort/solver.cpp:
```diff
@@ -0,0 +1,118 @@
+            // left-most char represents the bottom ball. Printing rows
+            // top-to-bottom and counting the top row as 1, the associated
+            // ball_index is MAX_CAPACITY minus the row number.
+            size_t ball_index = Tube::MAX_CAPACITY - row;
```
I see why it's public now, change it to a (static) getter

JustCallMeRay commented on src/ball_sort/solver.cpp:
```diff
@@ -0,0 +1,118 @@
+            bool has_balls_up_to_this_row =
+                tube.get_balls().size() > ball_index;
+            if (has_balls_up_to_this_row) {
+                std::cout << tube.get_balls().at(ball_index) << ' ';
```
Why the char in one place and the string lower down? Prefer consistency

JustCallMeRay commented on src/ball_sort/solver.cpp:
```diff
@@ -0,0 +1,118 @@
+        }
+        std::cout << '\n';
+    }
+    std::cout << '\n';
```
You will probably want to flush this here

JustCallMeRay commented on src/ball_sort/tube.cpp:
```diff
@@ -0,0 +1,73 @@
+
+auto Tube::is_one_colour() const -> bool
+{
+    std::unordered_set<char> ball_colours{};
+    for (const char &ball : m_balls) {
+        ball_colours.insert(ball);
+    }
```
slow, just check if they are all equal to the first one. (It's quicker to set(...) in python because C does that under the hood)

JustCallMeRay commented on tests/CMakeLists.txt:
```diff
@@ -0,0 +1,27 @@
+
+set(TEST_EXECUTABLE ball_sort_tests)
+
+FetchContent_Declare(
+  googletest
+  GIT_REPOSITORY https://github.com/google/googletest.git
+  GIT_TAG b479e7a3c161d7087113a05f8cb034b870313a55)
+
+FetchContent_MakeAvailable(googletest)
```
This makes me think the fetch content gets stuff from git them compiles it.

Great for small/macro heavy things but I don't want to compile GDAL every time I change the top level cmake file!
(Especially if my PM/technical lead doesn't like the globing)

So yeah conan/cpm is still better IMO but I see no reason to change it here.

JustCallMeRay commented on app/app.cpp:
Get some timeouts on this page to show off how quick the code is!

JustCallMeRay reviewed (changes requested):
I generally have my reference `&`s next to the type rather than the identifier, this is because the reference is part of the type not the name so I think it's more readable but it's a stylistic choice.

I think at least one function is `noexcept` but you haven't marked any as such (if wrongly marked as noexcept program crashes) , it's more useful for documentation than anything but I think you should still include it.

Looks really good, you are ready to start applying for places IMO

-- UPDATE: One of the files is really (in comparison) long, consider breaking it up a bit

JustCallMeRay commented on scripts.sh:
```diff
@@ -0,0 +1,31 @@
+
+case "$command" in
+"build-cmake")
+	cmake -B build
```
It'll print out the error, exit with the same code for pipelines though

DaveDangereux committed:
Rewrite is_one_colour to not use unordered_set

DaveDangereux committed:
Add exception if tube max capacity exceeded

DaveDangereux committed:
Add basic tests

DaveDangereux committed:
Add exception if empty tube is taken from

DaveDangereux committed:
Throw for illegal puzzles and moves

DaveDangereux committed:
Adopt left pointer alignment and add .clang-format file

DaveDangereux committed:
Throw for moving from out-of-bounds origin or destination

DaveDangereux committed:
Remove bin symlink and redirect binaries to top-level directory

DaveDangereux committed:
Make display of the solving process optional and allow playback speed adjustment

DaveDangereux committed:
Separate scripts into separate files and move to folder

DaveDangereux committed:
Update README and add license

DaveDangereux committed:
Add animated demo to README

DaveDangereux committed:
Add author to README

DaveDangereux committed:
Add known issues to README

DaveDangereux committed:
Make test.sh executable

DaveDangereux committed:
Make MAX_CAPACITY private and add static getter

DaveDangereux committed:
Add .DS_Store to .gitignore

DaveDangereux committed:
Add conan and make building tests optional

DaveDangereux committed:
Simplify exception messages with fmt

DaveDangereux committed:
Add nodiscard attributes and abandon trailing return type syntax

DaveDangereux committed:
Apply compiler flags conditionally for compatibility

DaveDangereux committed:
Return earlier in Puzzle::is_legal_move

DaveDangereux committed:
Add clang-tidy config file and fix many, many things

DaveDangereux committed:
Make all compiler options target-specific

DaveDangereux committed:
Replace .at calls with subscript operator

Closes DT-44

DaveDangereux committed:
Reserve size of vector in all make_tubes methods

In Puzzle::make_tubes, ensure that all vectors that are assigned to
iteratively have their size reserved beforehand.

Closes DT-41

DaveDangereux committed:
Move Move constructor and getters to header file

Closes DT-39

DaveDangereux committed:
Make pick_move and generate_filtered_moves private

In Solver, make pick_move and generate_filtered_moves (formerly
purge_redundant_moves) private methods. Update Solver tests to test
generate_filtered_moves via the public interface.

DaveDangereux committed:
Make Tube constructor take a string_view instead of a std::string

DaveDangereux committed:
Add default for serialised_state in Move constructor

Closes DT-32

DaveDangereux committed:
Overhaul scripts, cmake file formatting and add target_set_warnings function

DaveDangereux committed:
Add address and undefined behaviour sanitisers

Address and undefined behaviour sanitisers are on by default for debug
builds.

DaveDangereux committed:
Change Tube::get_balls() to return a std::string instead of a const reference

It was previously possible to return a dangling reference. Unfortunately
the address sanitiser was unable to catch this, but it has been
corrected regardless.

DaveDangereux committed:
Add Timer class and time solve method

DaveDangereux committed:
Add CodeRabbit and yamllint configs

DaveDangereux committed:
Add warning message for unsupported compilers in Warnings.cmake

DaveDangereux committed:
Remove safe.sh and replace with set -e in scripts

DaveDangereux committed:
Change assignments from get_serialised_state() to use value type

DaveDangereux committed:
Replace for loop with std::transform in Puzzle::make_tubes()

DaveDangereux committed:
Convert Solver class to solver namespace

DaveDangereux committed:
Enable CodeRabbit auto review on all branches

DaveDangereux committed:
Break up Puzzle implementation into smaller files

DaveDangereux committed:
Add guard against undo_move called with no history

DaveDangereux committed:
Remove redundant bool from Puzzle::validate_puzzle()

DaveDangereux committed:
Remove redundant print call from app CMakeLists.txt

DaveDangereux committed:
Reserve size for excluded_moves set

DaveDangereux committed:
Reserve for unordered_map in Puzzle::get_ball_tally()

DaveDangereux committed:
Add const to estimated_excluded_move_count in solver::solve

DaveDangereux committed:
Change solver::print_puzzle to optionally take a stream and a callback

DaveDangereux committed:
Put everything in a namespace

Closes DT-55

DaveDangereux committed:
Organise tests and methods

DaveDangereux committed:
Reword test suite names to match filenames

DaveDangereux committed:
Minor legibility tweaks and use regular snake case for all constant data-members

DaveDangereux committed:
Update README to reflect Solver class changing to ballsort::solver:: namespace

JustCallMeRay commented on .clang-tidy:
```diff
@@ -0,0 +1,16 @@
+---
```
I have a whole bunch of readability things in here normally to force my naming convention, I will send you an example (though it doesn't work for my preferred naming of lambdas)

D4n93r3ux commented:
@coderabbitai review

DaveDangereux committed:
Break up CMake sanitiser function and disable for tests due to weird bug

There seems to be a bug that causes the address sanitiser to be triggered by an empty test (from Google Test) or sometimes from a test with a test suite name that differs by one character from another test. Investigation is ongoing.

JustCallMeRay commented on .gitignore:
People normally have all .dll, .so, .obj, .o files in here, I use the default github provided one but you've marked the bin folder so should be fine. 

The intention should be to do a `git commit -am "message"` (or similar I may have got that wrong) without messing anything up or committing something you don't want to

JustCallMeRay commented on .yamllint.yaml:
This seems excessive to me but you do you!

JustCallMeRay commented on CMakeLists.txt:
```diff
@@ -0,0 +1,60 @@
+    elseif(CMAKE_CXX_COMPILER_ID MATCHES "Clang")
+        add_compile_options(-g)
+    elseif(CMAKE_CXX_COMPILER_ID MATCHES "GNU
+  ")
```
formatting is odd here?

JustCallMeRay commented on README.md:
```diff
@@ -0,0 +1,91 @@
+
+A `Puzzle` is instantiated by passing a vector of strings, with each character
+representing a ball colour:
+
```
I'm a bit iffy about suggesting the changing of source code, it's not a big deal here but I would not allow it in something that would be released as a product/lib (like at work)

Consider taking the puzzle string in as an input with an option to default to a solvable one

JustCallMeRay commented on README.md:
```diff
@@ -0,0 +1,91 @@
+be output to a top-level `bin` directory.
+
+Dependencies are currently managed with Conan, but support for alternative
+dependency managers will be added at a later date.
```
Add a `conan install .` suggestion here

JustCallMeRay commented on scripts/test.sh:
```diff
@@ -0,0 +1,8 @@
+rm -rf build
+conan install . -s build_type=Debug --build=missing -s compiler.cppstd=20
+cmake -DENABLE_SANITISE_ADDRESS=ON -DENABLE_SANITISE_UNDEFINED=ON -DENABLE_TESTING=ON --preset conan-debug
+cmake --build --preset conan-debug --target ball_sort_tests
```
I would expect `test.sh` to also run the tests. Change script or rename

JustCallMeRay commented on app/app.cpp:
```diff
@@ -0,0 +1,27 @@
+
+    char answer{};
+
+    std::cin >> answer;
```
I agree with the Code Rabbi - Tai

I also wouldn't use a char, if you do a loop where you ask for input again and the user wrote "yes" I think the next iteration gets populated with "e"

JustCallMeRay commented on cmake/Sanitisers.cmake:
```diff
@@ -0,0 +1,95 @@
+        message(
+            STATUS "[SANITISERS] Address sanitiser active for target: ${TARGET}"
+        )
+        target_add_compile_options(${TARGET} PRIVATE "/fsanitize=address")
```
You can actually use the dash instead of the slash here but you still need the elseif block so doesn't help much

JustCallMeRay commented on cmake/Sanitisers.cmake:
Normally I'd say there is a lot of boilerplate here and all the functions are visually similar but as it's Cmake I have no idea if the code I would like to write is possible. (A function that takes in sanitizer name, clang/gcc flag, msvc flag/error)

JustCallMeRay commented on cmake/Sanitisers.cmake:
```diff
@@ -0,0 +1,95 @@
+            WARNING "Undefined behaviour sanitiser not implemented for MSVC")
+
+    else()
+        message(WARNING "Compiler not supported for sanitisers")
```
These can be merged into one block

JustCallMeRay commented on cmake/Sanitisers.cmake:
```diff
@@ -0,0 +1,95 @@
+        message(WARNING "Thread sanitiser not implemented for MSVC")
+
+    else()
+        message(WARNING "Compiler not supported for sanitisers")
```
These can be merged into one block

JustCallMeRay commented on cmake/Warnings.cmake:
```diff
@@ -0,0 +1,95 @@
+    )
+
+    set(MSVC_WARNINGS
+        # Baseline
```
MSVC has security warnings called sdl checks

When a flag is given they become compile errors and are very rarely false positives.
I don't know what you should do with this information (sorry?)

JustCallMeRay commented on conanfile.py:
Personally I'd use the text file unless I am building a package/lib. 
Leave it though, shows you know the new things, maybe make it buildable with conan?

JustCallMeRay commented on include/ball_sort/exceptions/illegal_move_exception.hpp:
I generally don't like having a folder called exceptions, means I have to move through a bunch of files and folders if I want to edit something that mostly happens in one file. I prefer then near the code that uses them (so this would go near or in move.hpp

Also have a look at
https://en.cppreference.com/w/cpp/error/nested_exception

JustCallMeRay commented on include/ball_sort/exceptions/illegal_move_exception.hpp:
```diff
@@ -0,0 +1,14 @@
+    explicit IllegalMoveException(const std::string& message)
+        : std::runtime_error(message)
+    {}
+};
```
I tend to hold a string or stringstream, override `what`, and define a stream operator to append to the string on my exceptions

JustCallMeRay commented on include/ball_sort/exceptions/illegal_puzzle_exception.hpp:
See comments on move exception

JustCallMeRay commented on include/ball_sort/puzzle.hpp:
```diff
@@ -0,0 +1,51 @@
+
+    std::vector<Tube> m_initial_state;
+    std::vector<Tube> m_tubes;
+    std::vector<Move> m_move_history;
```
This can be default constructed (to empty) here

JustCallMeRay commented on include/ball_sort/solver.hpp:
```diff
@@ -0,0 +1,26 @@
+
+namespace ballsort::solver {
+
+using ClearCallback = std::function<void()>;
```
std::function does runtime checking so is slow(er than it should be), use a template as discussed and use a concept (std::invokeable) if required

JustCallMeRay commented on include/ball_sort/timer.hpp:
```diff
@@ -0,0 +1,22 @@
+    explicit Timer(const int precision = 3) : m_precision{precision} {};
+    void start();
+    void stop();
+    std::string get_time();
```
Should be no discard (and maybe const?)

JustCallMeRay commented on include/ball_sort/timer.hpp:
I have done timers before that are scope based:
A Stopwatch class creates a scope based timer (SBT) that contains a reference to the Stopwatch.
SBT starts the stopwatch at construction and stops it when destroyed.

You don't have to though this is fine!

JustCallMeRay commented on scripts/release.sh:
```diff
@@ -0,0 +1,8 @@
+
+rm -rf build
+conan install . -s build_type=Release --build=missing -s compiler.cppstd=20
+cmake -DENABLE_WARNINGS_AS_ERRORS=OFF --preset conan-release
```
You would normally want sanitizers on in release. 
Consider a debug, release and distribution if you want to make a distinction.
Debug only refers to what symbols are exposed and if a debugger can attach, release is often not treated a "fully built exportable binary" (less so with open source things as non-devy people need to build it and they get confused)

For example fuzz testing is done in release for the speed increase but also has sanitizers on

JustCallMeRay commented on scripts/test.sh:
```diff
@@ -0,0 +1,8 @@
+
+rm -rf build
+conan install . -s build_type=Debug --build=missing -s compiler.cppstd=20
+cmake -DENABLE_SANITISE_ADDRESS=ON -DENABLE_SANITISE_UNDEFINED=ON -DENABLE_TESTING=ON --preset conan-debug
```
You should be able to run tests in debug or release

JustCallMeRay commented on src/ball_sort/puzzle/puzzle_do_move.cpp:
```diff
@@ -0,0 +1,38 @@
+
+    const std::string current_state{get_serialised_state()};
+
+    if (m_visited_puzzle_states.contains(current_state)) {
```
If you are doing your own flat set make sure to give it a hint to start looking near the end (maybe 75% through instead of 50% in a binary search), 

I see why it was quicker and why a normal vector would be slow here (lots of checking from the start of the vector, all of which have multiple things to check)

I suggest profiling it first before listening to me again XD

JustCallMeRay commented on src/ball_sort/puzzle/puzzle_validate_puzzle.cpp:
```diff
@@ -0,0 +1,25 @@
+            throw IllegalPuzzleException(
+                "Puzzle must have four balls for each colour");
+        }
+    }
```
Looping over a map is normally slow

JustCallMeRay commented on src/ball_sort/solver.cpp:
```diff
@@ -0,0 +1,148 @@
+            const size_t milliseconds_per_move{5};
+            std::this_thread::sleep_for(
+                std::chrono::milliseconds(milliseconds_per_move));
+        }
```
Pause the timer if displaying then resume it again after (consider a scoped pause)

JustCallMeRay commented on src/ball_sort/solver.cpp:
```diff
@@ -0,0 +1,148 @@
+        if (origin.is_one_colour() && destination.is_empty()) { continue; }
+        if (excluded_moves.contains(move)) { continue; }
+
+        filtered_moves.emplace_back(move);
```
get a move itter of generate legal moves and (std::)move the move into filtered_moves

JustCallMeRay commented on src/ball_sort/solver.cpp:
```diff
@@ -0,0 +1,148 @@
+            // ball_index is MAX_CAPACITY minus the row number.
+            size_t ball_index = Tube::get_max_capacity() - row;
+
+            bool has_balls_up_to_this_row =
+                tube.get_balls().size() > ball_index;
```
Formatting looks odd here

JustCallMeRay commented on src/ball_sort/tube.cpp:
```diff
@@ -0,0 +1,84 @@
+    return m_balls;
+}
+
+std::string Tube::get_serialised_balls() const
+{
+    if (is_full()) { return m_balls; }
+
+    size_t gap_size{max_capacity - m_balls.size()};
+    return m_balls + std::string(gap_size, ' ');
+}
```
This doesn't quite go with my definition of serialisation, consider to_string (or a cast operator) instead

JustCallMeRay commented on tests/CMakeLists.txt:
```diff
@@ -0,0 +1,33 @@
+        ${ENABLE_WARNINGS_AS_ERRORS})
+endif()
+
+# The address sanitiser has been disabled for tests due to a peculiar bug that
+# can cause an empty test or tests with certain names to trigger the address
+# sanitiser
```
? Does this still happen in new version of clang?

JustCallMeRay commented on tests/puzzle/test_puzzle.cpp:
```diff
@@ -0,0 +1,40 @@
+
+TEST(PuzzleTest, LessThanMinimumTubesThrowsException)
+{
+    std::vector<std::string> tube_strings;
```
add a `{}` here

JustCallMeRay commented on include/ball_sort/move.hpp:
```diff
@@ -0,0 +1,41 @@
+struct Move {
+    friend std::hash<Move>;
+
+    explicit Move(const std::pair<size_t, size_t> move,
+                  std::string serialised_state = "")
+        : m_origin{move.first}, m_destination{move.second},
+          m_serialised_state{std::move(serialised_state)}
+    {}
```
You will probably want to have an overload that takes two `size_t`s to avoid the (minimal) overhead in constructing a pair

JustCallMeRay reviewed (changes requested):
When I assert an exception is thrown I like to check the error message has certain words to make sure it is relevant.

The comment about CVs from last time still applies! Looks great

DaveDangereux committed:
Sanitise multi-char user input

DaveDangereux committed:
Enable passing of command-line arguments

DaveDangereux committed:
Move to CPM and adopt canonical project structure

DaveDangereux committed:
Remove redundant debug print statement from add_test_targets.cmake

DaveDangereux committed:
Add ccache existence check to scripts

DaveDangereux committed:
Amend .clangd to point to correct compilation database location

DaveDangereux committed:
Do not offer to play back unsolvable puzzle

DaveDangereux committed:
Move unit tests next to relevant source files

DaveDangereux committed:
Incorporate GitHub templates into .gitignore

DaveDangereux committed:
Add nodiscard and const to Timer::get_time()

DaveDangereux committed:
Rename all serialised methods to as_string

DaveDangereux committed:
Add missing initialiser braces in puzzle.test.cpp

DaveDangereux committed:
Move exceptions near the code that uses them

DaveDangereux committed:
Add readability-identifier-naming checks to clang-tidy

DaveDangereux committed:
Amend scripts to build Debug and Release builds with optional tests

DaveDangereux committed:
Improve CMake sanitisers function

DaveDangereux committed:
Ensure builds work with clang and gcc and further refine sanitisers.cmake

DaveDangereux committed:
Add GitHub Actions (DT-93)

DaveDangereux committed:
Remove redundant print debugging line

DaveDangereux committed:
Remove old non-yaml .clangd file

DaveDangereux committed:
Update build instructions and restore to main README

DaveDangereux committed:
Correct typo in README

DaveDangereux committed:
Update README known issues and add planned improvements section

DaveDangereux committed:
Correct type on .clang-tidy config file

DaveDangereux committed:
Update GitHub URLs in README to use new username
